File: packages/core/src/tradewinds/core/temporal/knowledge_view.py
"""KnowledgeView — temporal filtering by ``knowledge_time``.

A plain wrapper class (not a pandas accessor, not a DataFrame subclass)
that exposes a ``dataframe()`` view filtered by ``knowledge_time <= as_of``.
This is the structural temporal-safety primitive the tradewinds research()
Mode 2 dispatch uses to render leakage-free training tables — any row
whose ``knowledge_time`` is later than the asserted as-of cutoff is
silently dropped from the view (and would be loud via :class:`LeakageDetector`).

Design constraints (CORE-07):

- Uses ``__slots__`` for memory predictability.
- Does NOT register a pandas accessor (verified by acceptance test).
- Validates input shape eagerly — raises :class:`SchemaValidationError`
  if ``knowledge_time`` is missing or not tz-aware UTC.

See ``docs/design.md`` §M (KnowledgeView semantics).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from tradewinds.core.exceptions import SchemaValidationError
from tradewinds.core.temporal.timepoint import TimePoint

if TYPE_CHECKING:
    import pandas as pd


__all__ = ["KnowledgeView"]


class KnowledgeView:
    """A filtered, knowledge-time-bounded view over a DataFrame.

    Construction validates the shape of the input DataFrame. After
    construction, :meth:`dataframe` returns a defensive copy of the rows
    where ``knowledge_time <= as_of``.
    """

    __slots__ = ("_df", "_as_of")

    def __init__(self, df: pd.DataFrame, as_of: TimePoint) -> None:
        import pandas as pd

        if "knowledge_time" not in df.columns:
            raise SchemaValidationError(
                "KnowledgeView requires 'knowledge_time' column",
                schema_id="<runtime>",
                violations=[{"column": "knowledge_time", "rule": "required"}],
            )
        if not isinstance(as_of, TimePoint):
            raise TypeError(f"as_of must be a TimePoint, got {type(as_of).__name__}")
        col = df["knowledge_time"]
        if not pd.api.types.is_datetime64_any_dtype(col):
            raise SchemaValidationError(
                "knowledge_time must be a datetime64 column",
                schema_id="<runtime>",
                violations=[{"column": "knowledge_time", "rule": "datetime_dtype"}],
            )
        if getattr(col.dt, "tz", None) is None:
            raise SchemaValidationError(
                "knowledge_time must be tz-aware UTC",
                schema_id="<runtime>",
                violations=[{"column": "knowledge_time", "rule": "tz_aware_utc"}],
            )
        self._df = df
        self._as_of = as_of

    def dataframe(self) -> pd.DataFrame:
        """Return a defensive copy of the rows where ``knowledge_time <= as_of``."""
        cutoff = self._as_of.to_utc()
        mask = self._df["knowledge_time"] <= cutoff
        return self._df.loc[mask].copy()

    @property
    def as_of(self) -> TimePoint:
        """The as-of cutoff supplied at construction."""
        return self._as_of

    def __repr__(self) -> str:
        return f"KnowledgeView(rows={len(self._df)}, " f"as_of={self._as_of.to_utc().isoformat()})"

File: packages-ts/core/src/temporal/knowledge-view.ts
// KnowledgeView — temporal filtering by `knowledge_time`.
//
// Mirrors `packages/core/src/tradewinds/core/temporal/knowledge_view.py`.
// A plain class (not an iterator subclass, not a DataFrame accessor) that
// filters its `rows()` output to entries where `knowledge_time <= asOf`.
//
// Construction validates the shape of EVERY row up-front: each
// `knowledge_time` must be a string that TimePoint accepts (rejecting naive
// / date-only / NaN inputs). Any failure throws SchemaValidationError with
// the violating row indices in `violations`.
//
// See docs/design.md §M for the leakage-free training-view semantics.

import { SchemaValidationError } from "../exceptions/index.js";
import { TimePoint } from "./timepoint.js";

/**
 * A filtered, knowledge-time-bounded view over an array of rows.
 *
 * @template Row — caller's row type. Must have a string `knowledge_time` field.
 */
export class KnowledgeView<Row extends { knowledge_time: string }> {
  readonly #rows: ReadonlyArray<Row>;
  readonly #asOf: TimePoint;
  readonly #asOfMs: number;

  constructor(rows: ReadonlyArray<Row>, asOf: TimePoint) {
    if (!(asOf instanceof TimePoint)) {
      throw new TypeError(`asOf must be a TimePoint; got ${typeof asOf}`);
    }
    // Eager per-row validation: every row's knowledge_time must parse as a
    // valid tz-aware ISO datetime. We piggyback on TimePoint's rejection
    // logic so the error vocabulary matches construction-time errors.
    const violations: Array<Record<string, unknown>> = [];
    for (let i = 0; i < rows.length; i++) {
      const r = rows[i];
      if (r == null || typeof r.knowledge_time !== "string") {
        violations.push({
          row_idx: i,
          rule: "knowledge_time_missing_or_wrong_type",
        });
        continue;
      }
      try {
        new TimePoint(r.knowledge_time);
      } catch (e) {
        violations.push({
          row_idx: i,
          rule: "knowledge_time_invalid",
          message: String(e),
        });
      }
    }
    if (violations.length > 0) {
      throw new SchemaValidationError(
        `KnowledgeView received ${violations.length} row(s) with invalid knowledge_time`,
        { schemaId: "<runtime>", violations },
      );
    }
    this.#rows = rows;
    this.#asOf = asOf;
    this.#asOfMs = asOf.toUTCDate().getTime();
  }

  /**
   * Return a freshly filtered array — only rows where
   * `knowledge_time <= asOf`. The returned array is a NEW reference each
   * call (defensive copy semantics), so callers can mutate it without
   * affecting subsequent calls.
   */
  rows(): ReadonlyArray<Row> {
    return this.#rows.filter((r) => Date.parse(r.knowledge_time) <= this.#asOfMs);
  }

  /** The as-of cutoff supplied at construction. */
  get asOf(): TimePoint {
    return this.#asOf;
  }
}
